modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -70,48 +161,126 @@ object GlobalL0Service {
+          .range(nextOrdinal.value.value, lastOrdinalCap + 1)
+          .map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o)))
+
+        type Success = Hashed[GlobalIncrementalSnapshot]
+        type Result = Chain[Success]
+        type Agg = (LazyList[SnapshotOrdinal], Result)
+        (ordinals, Chain.empty[Success]).tailRecM[F, Result] {
+          case (ordinal #:: nextOrdinals, snapshots) =>
+            l0GlobalSnapshotClient
+              .get(ordinal)(l0Peer)
+              .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
+              .map(s => (nextOrdinals, snapshots :+ s).asLeft[Result])
+              .handleErrorWith { e =>
+                logger
+                  .warn(e)(s"Failure pulling snapshot with ordinal=${ordinal.show}")
+                  .as(snapshots.asRight[Agg])
+              }
+
+          case (_, snapshots) => snapshots.asRight[Agg].pure[F]
+        }
+      }
+
+      private case class MajoritySnapshotData(ordinal: SnapshotOrdinal, hash: Hash)
+
+      private def getLatestMajoritySnapshotData(peerIds: NonEmptyList[PeerId]): F[MajoritySnapshotData] = {
+        val maybeData =
+          for {
+            peers <- OptionT.liftF(getL0Peers(peerIds))
+            majorityOrdinal <- OptionT(getMajorityOrdinal(peers))
+            majorityHash <- OptionT(getMajorityHash(peers, majorityOrdinal))
+          } yield MajoritySnapshotData(majorityOrdinal, majorityHash)
+
+        maybeData.getOrElseF(NoMajoritySnapshotData.raiseError[F, MajoritySnapshotData])
+      }
+
+      private def getMajorityHash(peers: NonEmptyList[L0Peer], ordinal: SnapshotOrdinal): F[Option[Hash]] =
+        peers.toList
+          .parTraverseN(numConcurrentQueries) { p =>
+            l0GlobalSnapshotClient
+              .getHash(ordinal)
+              .run(p)
+              .handleErrorWith(logger.warn(_)(s"Unable to obtain hash for majority peer ${p.show}").as(none[Hash]))
+          }
+          .map(_.flatten)
+          .flatTap(hashes => logger.debug(s"** Majority Hashes ${hashes.map(_.show)}"))
+          .map(selectMajorityHash)
+
+      private def pullVerifiedSnapshots(
+        nextOrdinal: SnapshotOrdinal,
+        msd: MajoritySnapshotData,
+        peerSet: NonEmptySet[L0Peer]
+      ): F[List[Hashed[GlobalIncrementalSnapshot]]] = {
+        type Agg = SortedSet[L0Peer]
+        type Result = List[Hashed[GlobalIncrementalSnapshot]]
+        Random
+          .shuffle(peerSet.toSortedSet)
+          .tailRecM[F, Result] { peers =>
+            peers.headOption.fold {
+              NoPeerAlignedWithMajority.raiseError[F, Either[Agg, Result]]
+            } { peer =>
+              pullSnapshots(peer, nextOrdinal, msd.ordinal).map { sc =>
+                Either.fromOption(
+                  sc.lastOption
+                    .filter(_.hash === msd.hash)
+                    .as(sc.toList)
+                    .filter(ss => ss.zip(ss.tail).forall(t => isNextSnapshot[GlobalIncrementalSnapshot](t._1, t._2))),
```

Comment:
@sdavidp please use `case (a, b) => ...` (and name a, b accordingly) instead of `_1`, `_2` which is ambiguous